js/tabState/task.js

```diff
@@ -1,93 +1,79 @@
-const tabPrototype = require("tabState/tab.js")
-const TabStack = require("tabRestore.js")
+const tabPrototype = require('tabState/tab.js')
+const TabStack = require('tabRestore.js')
 
-const taskPrototype = {
-  add: function (task, index) {
-    if (!task) {
-      task = {}
-    }
+class TaskList {
+  constructor () {
+    this.selected = null
+    this.tasks = [] // each task is {id, name, tabs: [], tabHistory: TabStack}
+  }
 
-    var newTask = {
+  add (task = {}, index) {
+    const newTask = {
       name: task.name || null,
       tabs: task.tabs || [],
       tabHistory: new TabStack(task.tabHistory),
-      id: task.id || String(getRandomId())
+      id: task.id || String(TaskList.getRandomId())
     }
 
     for (var key in tabPrototype) {
       newTask.tabs[key] = tabPrototype[key]
     }
 
     if (index) {
-      tabState.tasks.splice(index, 0, newTask)
+      this.tasks.splice(index, 0, newTask)
     } else {
-      tabState.tasks.push(newTask)
+      this.tasks.push(newTask)
     }
 
     return newTask.id
-  },
-  get: function (id) {
-    if (!id) {
-      return tabState.tasks
-    }
+  }
 
-    for (var i = 0; i < tabState.tasks.length; i++) {
-      if (tabState.tasks[i].id === id) {
-        return tabState.tasks[i]
-      }
+  getStringifyableState () {
+    return {
+      tasks: this.tasks,
+      selectedTask: this.selected
     }
-    return null
-  },
-  getTaskContainingTab: function (tabId) {
-    for (var i = 0; i < tabState.tasks.length; i++) {
-      if (tabState.tasks[i].tabs.has(tabId)) {
-        return tabState.tasks[i]
-      }
-    }
-    return null
-  },
-  getIndex: function (id) {
-    for (var i = 0; i < tabState.tasks.length; i++) {
-      if (tabState.tasks[i].id === id) {
-        return i
-      }
-    }
-    return -1
-  },
-  setSelected: function (id) {
-    tabState.selectedTask = id
-    window.currentTask = tasks.get(id)
+  }
+
+  get (id) {
+    return this.tasks.find(task => task.id == id) || null
```

Comment:
Array.find is like 10x slower than a for loop ([test](https://jsperf.com/find-vs-loop)). I'm not sure if this actually gets called enough that it would matter, but we might want to check how much of an impact this has.

Reply:
In that case we can keep the benefits of both ways by only changing the implementation of the TaskList::find () method, and use that instead of "this.tasks.find ()"

Reply:
Addressed in https://github.com/minbrowser/min/pull/622/commits/cd468e78c4f4a5460b36ee86f5dce1f8e0e7331a